pg_shard.c

```diff
@@ -329,6 +373,73 @@ PgShardPlanner(Query *query, int cursorOptions, ParamListInfo boundParams)
 
 
 /*
+ * SafeToPushDownAggregate returns true if GROUP BY clause includes the partition column.
+ * This function returns false if the total length of target list is greater than the
+ * total number of columns on the distributed table whose Oid is the second parameter
+ * of this function.
+ */
+static bool
+SafeToPushDownAggregate(Query *query, Oid distributedTableId)
+{
+	List *groupClauseList = query->groupClause;
+	List *targetList = query->targetList;
+	bool safeToPushDownAggregates = false;
+	Var *partitionColumn = PartitionColumn(distributedTableId);
+	char *finalAttributeName = get_attname(distributedTableId, list_length(targetList));
+
+	/*
+	 * standard_planner will be called on localQuery at PlanSequentialScan function.
+	 * In a sequential scan, there is an implicit check that the total number of unique
+	 * targetList elements cannot be greater than the total number of columns in the table.
+	 * Thus, it is not safe to push down aggregates for such cases.
+	 */
+	if (finalAttributeName == NULL)
+	{
+		safeToPushDownAggregates = false;
+	}
+	else
+	{
+		safeToPushDownAggregates = GroupedByColumn(groupClauseList, targetList,
+												   partitionColumn);
+	}
+
+	return safeToPushDownAggregates;
+}
+
+
+/*
+ * GroupedByColumn walks over group clauses in the given list, and checks if any
+ * of the group clauses is on the given column.
+ */
+static bool
+GroupedByColumn(List *groupClauseList, List *targetList, Var *column)
+{
+	bool groupedByColumn = false;
+	ListCell *groupClauseCell = NULL;
+
+	foreach(groupClauseCell, groupClauseList)
+	{
+		SortGroupClause *groupClause = (SortGroupClause *) lfirst(groupClauseCell);
+		TargetEntry *groupTargetEntry = get_sortgroupclause_tle(groupClause, targetList);
+
+		Expr *groupExpression = (Expr *) groupTargetEntry->expr;
+		if (IsA(groupExpression, Var))
```

Comment:
This function is directly taken from CitusDB. So, I didn't want to change it.
Should I add comment? Well, I don't know exactly why there is such a control.